pkg/summary/summary.go

```diff
@@ -49,119 +58,144 @@ type Summary struct {
 	Namespaces  []string            `json:"namespaces"`
 }
 
-// Client checks if VPA objects should be created or deleted
-// what? how does it do that? it checks it??
-type Client struct {
-	//changing these two to match the naming in here...but should it be consistent?
-	KubeClient    *kube.ClientInstance
-	KubeClientVPA *kube.VPAClientInstance
-}
+// Summarizer represents a source of generating a summary of VPAs
+type Summarizer struct {
+	Options
 
-var singleton *Client
+	// cached list of vpas
+	vpas []v1beta2.VerticalPodAutoscaler
+}
 
-// GetInstance returns a Client singleton
-func GetInstance() *Client {
-	if singleton == nil {
-		singleton = &Client{
-			KubeClient:    kube.GetInstance(),
-			KubeClientVPA: kube.GetVPAInstance(),
-		}
+// Returns a Summarizer for all goldilocks managed VPAs in all Namespaces
+func NewSummarizer(setters ...Option) *Summarizer {
+	opts := defaultOptions()
+	for _, setter := range setters {
+		setter(opts)
 	}
-	return singleton
-}
 
-// SetInstance sets the singleton using preconstructed k8s and vpa clients. Used for testing.
-func SetInstance(k8s *kube.ClientInstance, vpa *kube.VPAClientInstance) *Client {
-	singleton = &Client{
-		KubeClient:    k8s,
-		KubeClientVPA: vpa,
+	return &Summarizer{
+		Options: *opts,
 	}
-	return singleton
 }
 
-// Run creates a summary of the vpa info for all namespaces.
-func (client *Client) Run(vpaLabels map[string]string, excludeContainers string) (Summary, error) {
-	klog.V(3).Infof("Looking for VPAs with labels: %v", vpaLabels)
-
-	vpaListOptions := metav1.ListOptions{
-		LabelSelector: labels.Set(vpaLabels).String(),
-	}
+// Returns a Summarizer for a known list of VPAs
+func NewSummarizerForVPAs(vpas []v1beta2.VerticalPodAutoscaler, setters ...Option) *Summarizer {
+	summarizer := NewSummarizer(setters...)
 
-	vpas, err := client.KubeClientVPA.Client.AutoscalingV1beta2().VerticalPodAutoscalers("").List(vpaListOptions)
-	if err != nil {
-		klog.Error(err.Error())
-	}
-	klog.V(10).Infof("Found vpas: %v", vpas)
+	// set the cached vpas list directly
+	summarizer.vpas = vpas
 
-	summary, _ := GetInstance().constructSummary(vpas, excludeContainers)
-	return summary, nil
+	return summarizer
 }
 
-func (client *Client) constructSummary(vpas *v1beta2.VerticalPodAutoscalerList, excludeContainers string) (Summary, error) {
+func (s Summarizer) GetSummary() (Summary, error) {
```

Comment:
I think this should be a pointer
```suggestion
func (s Summarizer) GetSummary() (Summary, error) {
func (s *Summarizer) GetSummary() (Summary, error) {
```

Reply:
Why should this be a pointer receiver if the Summarizer struct shouldn't be modified during the function?

Reply:
Fair point, I usually like to keep my API of a struct consistent across methods, but you are correct that it doesn't necessarily have to. It may just be an idiosyncrasy of mine that isn't necessary.

The only downside I could think of is if anything in the future was added in this function that would modify the struct, it may not be obvious when things don't work as planned.

Reply:
I am actually leaning towards the approach you went with instead. Resolving this. Thanks!